### PR TITLE
Stack overflow

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Add stack overflow protection on all platforms.

On ARC, we fixup and runtime-test CONFIG_ARC_STACK_CHECKING which uses CPU features to detect stack overflow.

On all other arches, we have 100% software solution which writes a sentinel value which is checked on interrupt or context switch. It does not prevent corruption, and may not work in all cases, but should generally allow the user to know that stack bounds have been exceeded. This can be used on arches that don't have any HW features to help with stack checking, or lack RAM to enable them.

tests/kernel/fatal updated to test this.

See: https://jira.zephyrproject.org/browse/ZEP-1621

### CODE REVIEW

File: arch/arm/core/swap.S [hunk 1]
```diff
@@ -47,12 +47,18 @@ GDATA(_kernel)
...
-	push {lr}
-	bl _sys_k_event_logger_context_switch
-	pop {r0}
-	mov lr, r0
+    bl _sys_k_event_logger_context_switch
+#endif
+#ifdef CONFIG_STACK_SENTINEL
+    bl _check_stack_sentinel
+#endif
+    pop {lr}
```
Comment (andrewboie) [inline arch/arm/core/swap.S]: Huh, nrf51_pca10028, frdm_kl25z doesn't like "pop {lr}". I'll restore "pop {r0}; mov lr, r0"

Comment (pfalcon) [inline arch/arm/core/swap.S]: Perhaps ARMv6-M (Cortex-M0) limitation? If byte/clock-saving is important, perhaps a conditional would help.

File: subsys/debug/Kconfig [hunk 2]
```diff
@@ -23,6 +23,15 @@ config STACK_USAGE
...
 	on a per-function basis.
 
+config STACK_SENTINEL
+	bool "Enable stack sentinel"
+	select THREAD_STACK_INFO
+	default n
+	help
+	Store a magic value at the lowest addresses of a thread's stack.
+	Periodically check that this value is still present and kill the
+	thread gracefully if it isn't.
```
Comment (lpereira) [inline subsys/debug/Kconfig]: Might be a good idea to explain when the value is checked.

File: arch/nios2/core/fatal.c [hunk 3]
```diff
@@ -73,6 +73,11 @@ FUNC_NORETURN void _NanoFatalErrorHandler(unsigned int reason,
 		printk("***** Kernel Panic! *****\n");
 		break;
 
+#ifdef CONFIG_STACK_SENTINEL
```
Comment (lpereira) [inline arch/nios2/core/fatal.c]: We definitely need to unify _NanoFatalErrorHandler().

Comment (andrewboie) [inline arch/nios2/core/fatal.c]: https://jira.zephyrproject.org/browse/ZEP-2056

File: arch/arm/core/swap.S (hunk 1 above)
Comment (andrewboie) [inline arch/arm/core/swap.S]: Done

File: subsys/debug/Kconfig (hunk 2 above)
Comment (andrewboie) [inline subsys/debug/Kconfig]: Done

**Review by nashif — APPROVED** [review]

**Pull request merged**